File: ios/App/Sources/Features/Chats/ChatTitleGenerator.swift
import Foundation

/// Suggests short chat titles via **Lightweight Tasks** (Apple Intelligence
/// or a user-linked model). Falls back to a truncated first user message.
enum ChatTitleGenerator {
    static let defaultTitle = "New chat"
    static let maxTitleLength = 48

    /// Whether Apple Intelligence can run title generation right now.
    static var isOnDeviceAvailable: Bool {
        LightweightTaskRunner.appleFoundationAvailable
    }

    /// Build a display title from chat turns.
    static func suggestTitle(
        userMessages: [String],
        assistantMessages: [String] = []
    ) async -> String? {
        let heuristic = heuristicTitle(from: userMessages)
        guard let firstUser = userMessages
            .map({ $0.trimmingCharacters(in: .whitespacesAndNewlines) })
            .first(where: { !$0.isEmpty })
        else { return nil }

        let system = """
        You name chat conversations for a coding assistant app.
        Reply with ONLY a short title: 2 to 6 words.
        No quotes, no trailing punctuation, no emoji, no explanation.
        Capture the topic; prefer nouns over full sentences.
        """

        var body = "User: \(clip(firstUser, limit: 500))"
        for msg in userMessages.dropFirst().prefix(2) {
            body += "\nUser: \(clip(msg, limit: 200))"
        }
        for msg in assistantMessages.prefix(2) {
            body += "\nAssistant: \(clip(msg, limit: 200))"
        }
        body += "\n\nTitle:"

        if let generated = await LightweightTaskRunner.complete(
            system: system,
            user: body,
            maxTokens: 24
        ), let clean = sanitize(generated) {
            return clean
        }

        return heuristic
    }

    static func heuristicTitle(from userMessages: [String]) -> String? {
        guard let first = userMessages
            .map({ $0.trimmingCharacters(in: .whitespacesAndNewlines) })
            .first(where: { !$0.isEmpty })
        else { return nil }
        if first.count <= maxTitleLength { return first }
        return String(first.prefix(maxTitleLength - 1)) + "…"
    }

    static func sanitize(_ raw: String) -> String? {
        var text = ThinkingExtractor.plainVisibleText(from: raw)
        guard !text.isEmpty else { return nil }

        if let newline = text.firstIndex(of: "\n") {
            text = String(text[..<newline]).trimmingCharacters(in: .whitespacesAndNewlines)
        }

        let wrappers: [Character] = ["\"", "'", "`", "*", "“", "”", "‘", "’"]
        while let first = text.first, let last = text.last,
              wrappers.contains(first), wrappers.contains(last),
              text.count > 2 {
            text = String(text.dropFirst().dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
        }

        if text.hasSuffix("."), text.count <= 40, !text.dropLast().contains(".") {
            text = String(text.dropLast())
        }

        guard !text.isEmpty else { return nil }
        if text.count <= maxTitleLength { return text }
        return String(text.prefix(maxTitleLength - 1)) + "…"
    }

    private static func clip(_ text: String, limit: Int) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.count <= limit { return trimmed }
        return String(trimmed.prefix(limit)) + "…"
    }
}

File: android/app/src/main/kotlin/app/roamsocket/android/ui/chats/ChatTitleGenerator.kt
package app.roamsocket.android.ui.chats

import app.roamsocket.android.AppContainer
import app.roamsocket.android.ui.lightweight.LightweightTaskRunner
import app.roamsocket.core.chats.PersistedChatMessage

/**
 * Suggests a short display title for a chat. Mirrors the iOS
 * `Chats/ChatTitleGenerator.swift` surface:
 *
 * 1. If the user has a linked lightweight model, ask it for a
 *    short title (1-6 words) and sanitize the result.
 * 2. Otherwise, fall back to the heuristic "first 6 words of the
 *    first user message".
 *
 * Never overwrites a user-edited title; the caller is responsible
 * for that check.
 */
object ChatTitleGenerator {

    const val MAX_TITLE_LENGTH: Int = 48

    private const val SYSTEM_PROMPT =
        "You title chat conversations. Reply with a short, descriptive " +
            "title of 1-6 words. No quotes, no trailing punctuation, no " +
            "explanation. Output only the title."

    /**
     * Suggest a title for the chat whose first user message is
     * `messages.firstOrNull { role == USER }`. Returns
     * `currentTitle` unchanged when the chat has no user message
     * yet.
     */
    suspend fun suggestTitle(
        container: AppContainer,
        messages: List<PersistedChatMessage>,
        currentTitle: String,
    ): String {
        val firstUser = messages.firstOrNull { it.role == PersistedChatMessage.Role.USER }
            ?: return currentTitle
        val firstText = firstUser.content.trim()
        if (firstText.isEmpty()) return currentTitle

        val lightweight = LightweightTaskRunner.complete(
            container = container,
            system = SYSTEM_PROMPT,
            user = firstText,
            maxTokens = 16,
        )
        if (!lightweight.isNullOrBlank()) {
            return sanitize(lightweight)
        }
        return heuristic(firstText)
    }

    /** First 6 words of [text], capped at [MAX_TITLE_LENGTH]. */
    fun heuristic(text: String): String {
        val cleaned = text.trim().replace(Regex("\\s+"), " ")
        if (cleaned.isEmpty()) return ""
        val words = cleaned.split(' ').take(6).joinToString(" ")
        return if (words.length <= MAX_TITLE_LENGTH) {
            words
        } else {
            // Reserve one slot for the ellipsis so the final string
            // never exceeds [MAX_TITLE_LENGTH] (matches iOS
            // `prefix(maxTitleLength - 1) + "…"`).
            words.substring(0, MAX_TITLE_LENGTH - 1).trimEnd { it == ' ' || it == ',' } + "\u2026"
        }
    }

    /**
     * Strip common wrappers and prefixes an LLM might add (quotes,
     * leading "Title:", trailing period) and cap at
     * [MAX_TITLE_LENGTH].
     */
    fun sanitize(raw: String): String {
        var text = raw.trim()
        if (text.isEmpty()) return ""

        val newline = text.indexOf('\n')
        if (newline >= 0) text = text.substring(0, newline).trim()

        val wrappers = charArrayOf('"', '\'', '`', '*', '\u201C', '\u201D', '\u2018', '\u2019')
        while (
            text.length > 2 &&
            wrappers.contains(text.first()) &&
            wrappers.contains(text.last())
        ) {
            text = text.substring(1, text.length - 1).trim()
        }

        for (prefix in listOf("Title:", "Name:", "Chat:")) {
            if (text.startsWith(prefix, ignoreCase = true)) {
                text = text.substring(prefix.length).trim()
            }
        }

        if (text.endsWith('.') && text.length <= 36 && !text.dropLast(1).contains('.')) {
            text = text.dropLast(1)
        }

        if (text.isEmpty()) return ""
        if (text.length <= MAX_TITLE_LENGTH) return text
        return text.take(MAX_TITLE_LENGTH - 1) + "\u2026"
    }
}
